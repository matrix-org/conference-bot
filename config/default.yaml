# Where the homeserver's Client-Server API is located. Typically this
# is where clients would be connecting to in order to send messages.
homeserverUrl: "https://matrix-client.matrix.org"

# An access token for the bot to use. This should be the as_token from the
# appservice registration.
accessToken: "YOUR_TOKEN_HERE"

# Set this to the desired name for the bot. The account must already be
# registered and accessible as an impersonatable user ID by the accessToken.
userId: "@bot:example.org"

# Location on disk for where to store various bot information.
dataPath: "storage"

# The ID of the room to use as a management room. The bot will only
# accept commands in here.
managementRoom: "!room:example.org"

# The identity server to use to send email invites to any participants. The
# bot will accept all terms of service automatically - please ensure that it
# has legal capability to do so before running the bot in this way.
idServerDomain: "vector.im"

# The brand to supply to the identity server when performing actions which may
# result in emails or other communications to the user.
idServerBrand: "vector-im"

# This is the user ID of the moderator for all public-facing rooms the bot creates.
# Note that this user will be granted power level 100 (the highest) in every room
# and be invited.
moderatorUserId: "@moderator:example.org"

# Settings for how the bot should represent livestreams.
livestream:
  # The template for livestreams in auditorium rooms
  # Available variables:
  #  id  - The room ID/name (eg: "D.collab")
  #  sId - The room ID/name, but lowercase and only alphanumeric characters
  #        are included (eg: "dcollab")
  auditoriumUrl: "https://stream.example.org/conference/hls/{id}.m3u8"

  # The template for livestreams in talk rooms
  # Available variables:
  #  slug  - The shorthand version of the talk name (eg: "matrix-p2p")
  #  audId - The auditorium ID the talk is held in (eg: "D.collab")
  #  jitsi - The Jitsi conference name in lowercase.
  talkUrl: "https://stream.example.org/conference-talks/hls/{slug}.m3u8"

  # The template for livestreams in random rooms.
  # Available variables:
  #  jitsi - The Jitsi conference name in lowercase.
  hybridUrl: "https://stream.example.org/conference-hybrids/hls/{jitsi}.m3u8"

  # The URL template for the schedule widget.
  # Available variables:
  #  audId - The auditorium ID the talk is held in (eg: "D.collab")
  scheduleUrl: "https://example.org/?theme=$theme&room={audId}"

  # The domain of the Jitsi server to use. It will be assumed to be using
  # https://github.com/matrix-org/prosody-mod-auth-matrix-user-verification
  jitsiDomain: "jitsi.riot.im"

  # An MSC URI to use as the avatar for widgets generated by the bot.
  widgetAvatar: "mxc://example.org/mediaid"

  # Options for the /onpublish endpoint used to redirect streams from Jitsi
  # to an RTMP location. Query string arguments are:
  #  auth=secret   - See below
  #  conf=aabbcc   - The Jitsi conference name. This should be a base32 string
  #                  when using prosody-mod-auth-matrix-user-verification as above.
  # If no mapping is found, 200 OK is returned for compatibility with nginx.
  onpublish:
    # The hostname to use in the rtmpPrefix template. Some variables are supported
    # here:
    #  squishedAudId  - The auditorium ID with all non-alphanumeric characters
    #                   removed, and lowercase.
    # Note: The hostname will be resolved to an IP address before used as a hostname
    # for compatibility with nginx.
    rtmpHostnameTemplate: "{squishedAudId}.example.org"

    # The URL to redirect to, including scheme.
    # Some variables are supported here:
    #  hostname    - The hostname as generated/resolved by rtmpHostnameTemplate
    #  saltedHash  - The salted hash generated by the bot to secure the stream with.
    rtmpUrlTemplate: "rtmp://{hostname}/stream/{saltedHash}"

    # The salt to use in the hash. The hash will be "event_id.secret" where
    # "event_id" is the penta event ID.
    #
    # The value which must also be supplied as ?auth to /onpublish in order to get
    # a proper 302 redirect. When the auth mismatches a 200 status code will be
    # returned to imply there is no mapping.
    salt: "change me"

# Configuration for the webserver (used to serve widgets for the conference's
# rooms).
webserver:
  # The address to bind to. If using Docker, this should be all interfaces and
  # instead managed through port mapping.
  address: '0.0.0.0'

  # The port to bind the webserver on.
  port: 8080

  # The public URL where the bot can expect to be able to serve widgets from.
  # Note that this URL will be put into the Matrix room state.
  publicBaseUrl: "https://conferences.example.org"

  # Additional assets may be served from `${publicBaseUrl}/assets/*` which are
  # available at this path.
  additionalAssetsPath: "/data/public"

# Information about the conference
conference:
  # The ID of the conference to call this one. Data from other conferences
  # will be ignored (such as past conferences). This will also be used to create
  # the Matrix Space.
  id: "fosdem2021"

  # The name of the conference.
  name: "FOSDEM 2021"

  # The URL to the XML which is updated with conference information.
  # This is read and parsed by the bot during the early stages of
  # setting up the conference.
  pentabarfDefinition: "https://fosdem.org/2021/schedule/xml"

  # The timezone that the the bot's database is operating off of.
  timezone: "Europe/Brussels"

  # The various support rooms the bot supports. They will be created if they
  # haven't already.
  supportRooms:
    speakers: "#speakers:example.org"
    specialInterest: "#stands:example.org"
    coordinators: "#coordinators:example.org"

  # How far in the future the bot should be looking before considering something "finalized".
  # Should never be less than 5 minutes.
  lookaheadMinutes: 5

  # Connection information to get data about the conference during the conference.
  # This can be readonly. Currently the bot only supports postgresql.
  database:
    host: "localhost"
    port: 5432
    username: "db_username"
    password: "db_password"
    database: "your_db"
    sslmode: "require"

    # The table/view for the bot to access for information about people involved in the conference.
    # Must at least have the following columns:
    #  event_id         - TEXT(like) - The penta ID for the talk (event) associated with
    #                                  the person, if relevant. Null if a coordinator.
    #  person_id        - TEXT(like) - The penta ID for the person.
    #  event_role       - TEXT(like) - One of [speaker, coordinator, host]. See roles later on.
    #  name             - TEXT(like) - The full name, or otherwise useful name, for the person.
    #  email            - TEXT(like) - The preferred email address for the person.
    #  matrix_id        - TEXT(like) - If known, the Matrix User ID for the person, otherwise null.
    #  conference_room  - TEXT(like) - The relevant room. This should match the event_id or be
    #                                  the room where coordinators are assigned.
    #
    # People may be assigned multiple roles with multiple rows.
    #
    # Roles:
    #  speaker      - A speaker for event_id. This person will get moderator in their talk's
    #                 room, and invited to the auditorium backstage room. They will be required
    #                 to check in before their talk starts.
    #
    #  host         - Someone who is moderating the talk itself in collaboration with the speakers
    #                 for the event_id. One host must check in before the talk starts, otherwise
    #                 the issue will be raised to the management room and coordinators will be
    #                 asked to take over. Hosts will get moderator in their talk's room, and be
    #                 invited to the public auditorium room, backstage room, and talk room.
    #
    #  coordinator  - These are typically people responsible for scheduling the auditorium's
    #                 talks. They'll get moderator in the auditorium room, backstage room, and all
    #                 talk rooms for their auditorium, though will only be invited to the auditorium
    #                 and backstage room (and the talk rooms on-demand if needed to fill in for a
    #                 missing host).
    #
    # The bot will ignore unknown events, rooms, and roles.
    #
    # CAUTION: Although the bot uses parameterized queries, it is unable to use the table name as
    # a parameter. As such, this particular config value is vulnerable to SQL injection. Seeing as
    # how you (the bot's admin) are the one entering it: don't do that to yourself.
    tblPeople: "view_matrix_bot_export_people"

    # The table/view for the bot to access for information about the conference schedule.
    # Must at least have the following columns:
    #  event_id            - TEXT(like) - The penta ID for the talk (event).
    #  conference_room     - TEXT(like) - The relevant room. This should match the event_id.
    #  start_datetime      - TIMESTAMP WITHOUT TIME ZONE - The start time of the talk. The timezone
    #                        is determined by the bot using the timezone elsewhere in this config.
    #  duration            - INTERVAL   - How long the talk is meant to last (start to end).
    #  presentation_length - INTERVAL   - The portion of `duration` where the speaker is presenting.
    #                                     The remainder is assumed to be Q&A.
    #  prerecorded         - BOOLEAN    - True if the presentation is prerecorded. When false, the
    #                                     bot will assume that presentation_length is zero and switch
    #                                     directly into "hallway" mode at start_datetime.
    #
    # The bot will ignore unknown talks. Any if the start_datetime is NULL then the bot will ignore
    # that record. If the INTERVAL fields are NULL, the bot will assume zero.
    #
    # Note: the bot is capable of running both tblSchedule and tblPeople off the same table/view.
    # It will use DISTINCT where it needs to in order to get accurate information.
    #
    # CAUTION: Although the bot uses parameterized queries, it is unable to use the table name as
    # a parameter. As such, this particular config value is vulnerable to SQL injection. Seeing as
    # how you (the bot's admin) are the one entering it: don't do that to yourself.
    tblSchedule: "view_matrix_bot_export_schedule"

    # The time in seconds added to the presentation_length to account for preroll material such
    # as sponsor segments. If this is built into the table already then set this to zero.
    scheduleBufferSeconds: 30

  # Various prefixes used by the bot when parsing information.
  prefixes:
    # The prefixes for the rooms listed in the pentabarf definition which
    # describe the rooms that are 'auditorium' rooms. The prefixes will be
    # trimmed when naming the room (unless an override is present).
    auditoriumRooms: ["D."]

    # The prefixes for the rooms listed in the pentabarf definition which
    # describe the rooms that are 'special interest' rooms. The prefixes will
    # be trimmed when naming the room (unless an override is present).
    interestRooms: ["S.", "B."]

    # Simple mapping of room IDs to chosen names. The names should not contain
    # spaces or any complex characters. This is typically used to resolve naming
    # conflicts after trimming the prefix.
    nameOverrides:
      #"D.example": "example-room"

    # The prefix to apply to all generated room aliases.
    aliases: ""

    # The alias suffixes to apply for certain prefixes of rooms
    suffixes:
      #"S.": "-stand"

    # The display name suffixes to apply for certain prefixes of rooms
    displayNameSuffixes:
      #"S.": " stand"

  # A mapping of pre-existing rooms to re-use as interest rooms.
  existingInterestRooms:
    #"I.infodesk": "#infodesk:example.org"

  # The subspaces to create under the main conference space
  subspaces:
    #stands:
    #  displayName: Stands
    #  alias: stands
    #  # The prefixes of rooms which belong in the subspace
    #  prefixes: ["S."]

# Options related to the IRC bridge.
ircBridge:
  # The nick to use when taking ownership of channels
  botNick: 'confbot'
  botPassword: 'a_secret_password'

  # The server name of the IRCd
  serverName: 'localhost'

  # The port of the IRCd
  port: 6667

  # The userId of the irc bridge
  botUserId: '@appservice-irc:example.com'

  # The allowed channel prefix of the bot
  channelPrefix: '#conference-'
