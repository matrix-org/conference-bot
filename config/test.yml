# We don't set any config for tests currently, but node-config complains if we don't have this file.

