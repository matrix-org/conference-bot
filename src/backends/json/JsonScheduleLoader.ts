import { RoomKind } from "../../models/room_kinds";
import { IAuditorium, IConference, IInterestRoom, IPerson, ITalk, Role } from "../../models/schedule";
import { JSONSchedule, JSONSpeaker, JSONStream, JSONTalk } from "./jsontypes/JsonSchedule.schema";
import * as moment from "moment";
import { AuditoriumId, InterestId, TalkId } from "../IScheduleBackend";

/**
 * Loader and holder for JSON schedules.
 * Maybe a misnomer; perhaps 'JsonSchedule' is better? :/
 */
export class JsonScheduleLoader {
    public readonly conference: IConference;
    public readonly auditoriums: Map<AuditoriumId, IAuditorium>;
    public readonly talks: Map<TalkId, ITalk>;
    public readonly interestRooms: Map<InterestId, IInterestRoom>;
    public readonly conferenceId: string;
    //public readonly speakers: IPerson[];

    constructor(jsonDesc: object) {
        // TODO: Validate and give errors. Assuming it's correct is a bit cheeky.
        const jsonSchedule = jsonDesc as JSONSchedule;

        this.auditoriums = new Map();
        this.talks = new Map();

        for (let aud of jsonSchedule.streams) {
            const auditorium = this.convertAuditorium(aud);
            if (this.auditoriums.has(auditorium.id)) {
                throw `Conflict in auditorium ID «${auditorium.id}»!`;
            }
            this.auditoriums.set(auditorium.id, auditorium);

            for (let talk of auditorium.talks.values()) {
                if (this.talks.has(talk.id)) {
                    const conflictingTalk = this.talks.get(talk.id)!;
                    throw `Talk ID ${talk.id} is not unique — occupied by both «${talk.title}» and «${conflictingTalk.title}»!`;
                }
                this.talks.set(talk.id, talk);
            }
        }

        // TODO: Interest rooms are currently not supported by the JSON schedule backend.
        this.interestRooms = new Map();

        this.conference = {
            title: jsonSchedule.title,
            auditoriums: Array.from(this.auditoriums.values()),
            interestRooms: Array.from(this.interestRooms.values())
        };
    }

    private convertSpeaker(speaker: JSONSpeaker): IPerson {
        return {
            id: this.slugify(speaker.display_name),
            name: speaker.display_name,
            matrix_id: speaker.matrix_id,
            email: speaker.email,
            role: Role.Speaker
        };
    }

    private convertTalk(talk: JSONTalk, auditoriumId: string): ITalk {
        const startMoment = moment.utc(talk.start, moment.ISO_8601, true);
        const endMoment = moment.utc(talk.end, moment.ISO_8601, true);

        return {
            id: talk.id.toString(), // TODO We have numbers on the rhs and a string on the lhs.
            title: talk.title,
            subtitle: talk.description, // TODO is this valid?
            slug: this.slugify(talk.title),

            auditoriumId,
            prerecorded: true, // TODO
            qa_startTime: null, // TODO
            livestream_endTime: endMoment.valueOf(), // TODO is this right?
            speakers: talk.speakers.map(speaker => this.convertSpeaker(speaker)),
            track: "", // TODO we have multiple of them!!!

            // Must .clone() here because .startOf() mutates the moment(!)
            dateTs: startMoment.clone().startOf("day").valueOf(),
            startTime: startMoment.valueOf(),
            endTime: endMoment.valueOf(),
        };
    }

    private convertAuditorium(stream: JSONStream): IAuditorium {
        const auditoriumId = this.slugify(stream.stream_name);

        const talks: Map<TalkId, ITalk> = new Map();

        for (let unconvTalk of stream.talks) {
            const talk = this.convertTalk(unconvTalk, auditoriumId);
            if (talks.has(talk.id)) {
                const conflictingTalk = talks.get(talk.id)!;
                throw `Talk ID ${talk.id} is not unique — occupied by both «${talk.title}» and «${conflictingTalk.title}»!`;
            }
            talks.set(talk.id, talk);
        }

        return {
            id: auditoriumId,
            slug: this.slugify(stream.stream_name),
            name: stream.stream_name,
            kind: RoomKind.Auditorium,
            talks,
            // TODO Support physical auditoriums in the JSON schedule backend
            isPhysical: false,
        };
    }

    /**
     * Convert a string to something that is usable as a slug / ID.
     * The result only contains the characters in [a-z0-9-_].
     */
    private slugify(input: string): string {
        return input.toLowerCase().replace(/[^0-9a-z-_]+/g, "_");
    }
}