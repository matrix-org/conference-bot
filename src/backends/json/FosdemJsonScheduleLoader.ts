import { RoomKind } from "../../models/room_kinds";
import { IAuditorium, IConference, IInterestRoom, IPerson, ITalk, Role } from "../../models/schedule";
import { FOSDEMSpecificJSONSchedule, FOSDEMPerson, FOSDEMTrack, FOSDEMTalk } from "./jsontypes/FosdemJsonSchedule.schema";
import * as moment from "moment";
import { AuditoriumId, InterestId, TalkId } from "../IScheduleBackend";
import { IConfig } from "../../config";

/**
 * Loader and holder for FOSDEM-specific JSON schedules, acquired from the
 * custom `/p/matrix` endpoint on the Pretalx instance.
 */
export class FosdemJsonScheduleLoader {
    public readonly conference: IConference;
    public readonly auditoriums: Map<AuditoriumId, IAuditorium>;
    public readonly talks: Map<TalkId, ITalk>;
    public readonly interestRooms: Map<InterestId, IInterestRoom>;
    public readonly conferenceId: string;

    constructor(jsonDesc: object, globalConfig: IConfig) {
        // TODO: Validate and give errors. Assuming it's correct is a bit cheeky.
        const jsonSchedule = jsonDesc as FOSDEMSpecificJSONSchedule;

        this.auditoriums = new Map();

        for (let rawTrack of jsonSchedule.tracks) {
            // Tracks are now (since 2025) mapped 1:1 to auditoria
            const auditorium = this.convertAuditorium(rawTrack);
            if (this.auditoriums.has(auditorium.id)) {
                throw `Conflict in auditorium ID «${auditorium.id}»!`;
            }
            this.auditoriums.set(auditorium.id, auditorium);
        }

        this.talks = new Map();

        for (let rawTalk of jsonSchedule.talks) {
            const talk = this.convertTalk(rawTalk);
            if (this.talks.has(talk.id)) {
                const conflictingTalk = this.talks.get(talk.id)!;
                throw `Talk ID ${talk.id} is not unique — occupied by both «${talk.title}» and «${conflictingTalk.title}»!`;
            }
            const auditorium = this.auditoriums.get(talk.auditoriumId);
            if (!auditorium) {
                throw `Talk ID ${talk.id} relies on non-existent auditorium ${talk.auditoriumId}`;
            }
            auditorium.talks.set(talk.id, talk);
            this.talks.set(talk.id, talk);
        }

        // TODO: Interest rooms are currently not supported by the JSON schedule backend.
        this.interestRooms = new Map();

        this.conference = {
            title: globalConfig.conference.name,
            auditoriums: Array.from(this.auditoriums.values()),
            interestRooms: Array.from(this.interestRooms.values())
        };
    }

    private convertPerson(person: FOSDEMPerson): IPerson {
        if (! Object.values<string>(Role).includes(person.event_role)) {
            throw new Error("unknown role: " + person.event_role);
        }
        return {
            id: person.person_id.toString(),
            name: person.name,
            matrix_id: person.matrix_id,
            email: person.email,
            // safety: checked above
            role: person.event_role as Role,
        };
    }

    private convertTalk(talk: FOSDEMTalk): ITalk {
        const auditoriumId = talk.track.id.toString();
        const startMoment = moment.utc(talk.start_datetime, moment.ISO_8601, true);
        const endMoment = startMoment.add(talk.duration, "minutes");

        return {
            id: talk.event_id.toString(),
            title: talk.title,

            // Pretalx does not support this concept. FOSDEM 2024 ran with empty strings. From 2025 we hardcode this as empty for now.
            subtitle: "",

            auditoriumId,

            // Hardcoded: all talks are now live from FOSDEM 2025
            prerecorded: false,

            // This is sketchy, but the QA start-time is not applicable except to prerecorded talks.
            // Even then, it's not clear why it would be different from the end of the talk?
            // This is overall a messy concept, but the only thing that matters now is whether this is
            // null (Q&A disabled) or non-null (Q&A enabled, with reminder 5 minutes before the end of the talk slot).
            // TODO overhaul replace with a boolean instead...?
            qa_startTime: 0,

            // Since the talks are not pre-recorded, the livestream is considered ended when the event ends.
            livestream_endTime: endMoment.valueOf(),

            speakers: talk.persons.map(person => this.convertPerson(person)),

            // Must .clone() here because .startOf() mutates the moment(!)
            dateTs: startMoment.clone().startOf("day").valueOf(),
            startTime: startMoment.valueOf(),
            endTime: endMoment.valueOf(),
        };
    }

    private convertAuditorium(track: FOSDEMTrack): IAuditorium {
        const extraPeople: IPerson[] = track.managers.map(person => this.convertPerson(person));
        return {
            id: track.id.toString(),
            slug: track.slug,
            name: track.name,
            kind: RoomKind.Auditorium,
            // This will be populated afterwards
            talks: new Map(),
            // Hardcoded: FOSDEM is always physical now.
            isPhysical: true,
            extraPeople,
            trackType: track.type,
        };
    }
}
