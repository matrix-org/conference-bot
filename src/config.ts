/*
Copyright 2020, 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as config from "config";
import { MatrixClient } from "matrix-bot-sdk";
import { Conference } from "./Conference";
import { IRCBridge, IRCBridgeOpts } from "./IRCBridge";
import { Scheduler } from "./Scheduler";
import { CheckInMap } from "./CheckInMap";

interface IConfig {
    homeserverUrl: string;
    accessToken: string;
    userId: string;
    dataPath: string;
    managementRoom: string;
    idServerDomain: string;
    idServerBrand: string;
    moderatorUserId: string;
    livestream: {
        auditoriumUrl: string;
        talkUrl: string;
        hybridUrl: string;
        scheduleUrl: string;
        jitsiDomain: string;
        widgetAvatar: string;
        onpublish: {
            rtmpHostnameTemplate: string;
            rtmpUrlTemplate: string;
            salt: string;
        };
    };
    webserver: {
        address: string;
        port: number;
        publicBaseUrl: string;
        additionalAssetsPath: string;
    };
    conference: {
        id: string;
        name: string;
        pentabarfDefinition: string;
        timezone: string;
        lookaheadMinutes: number;
        supportRooms: {
            speakers: string;
            coordinators: string;
            specialInterest: string;
        };
        prefixes: {
            auditoriumRooms: string[];
            interestRooms: string[];
            aliases: string;
            suffixes: {
                [prefix: string]: string;
            };
            displayNameSuffixes: {
                [prefix: string]: string;
            };
            nameOverrides: {
                [auditoriumId: string]: string;
            };
        };
        database: {
            host: string;
            port: number;
            username: string;
            password: string;
            database: string;
            sslmode: string;
            tblPeople: string;
            tblSchedule: string;
            scheduleBufferSeconds: number;
        };
    };
    ircBridge: IRCBridgeOpts;

    RUNTIME: {
        client: MatrixClient;
        conference: Conference;
        scheduler: Scheduler;
        ircBridge: IRCBridge;
        checkins: CheckInMap;
    };
}

export default <IConfig>config;
