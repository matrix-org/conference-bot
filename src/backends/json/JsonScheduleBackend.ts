import { readFile } from "fs";
import { IJsonScheduleBackendConfig } from "../../config";
import { IConference, ITalk, IAuditorium, IInterestRoom } from "../../models/schedule";
import { AuditoriumId, InterestId, IScheduleBackend, TalkId } from "../IScheduleBackend";
import { JsonScheduleLoader } from "./JsonScheduleLoader";
import * as fetch from "node-fetch";

/**
 * Reads a JSON file from disk.
 */
 function readJsonFileAsync(path: string): Promise<object> {
    return new Promise((resolve, reject) => {
        readFile(path, {}, (err, buf: string) => {
            if (err) {
                reject(err);
            } else {
                try {
                    resolve(JSON.parse(buf));
                } catch (err) {
                    reject(err);
                }
            }
        })
    });
}

export class JsonScheduleBackend implements IScheduleBackend {
    constructor(private loader: JsonScheduleLoader, private cfg: IJsonScheduleBackendConfig) {

    }

    private static async loadConferenceFromCfg(cfg: IJsonScheduleBackendConfig): Promise<JsonScheduleLoader> {
        let jsonDesc;
        if (cfg.scheduleDefinition.startsWith("http")) {
            // Fetch the JSON track over the network
            jsonDesc = await fetch(cfg.scheduleDefinition).then(r => r.json());
        } else {
            // Load the JSON from disk
            jsonDesc = await readJsonFileAsync(cfg.scheduleDefinition);
        }

        return new JsonScheduleLoader(jsonDesc);
    }

    static async new(cfg: IJsonScheduleBackendConfig): Promise<JsonScheduleBackend> {
        const loader = await JsonScheduleBackend.loadConferenceFromCfg(cfg);
        return new JsonScheduleBackend(loader, cfg);
    }

    async refresh(): Promise<void> {
        this.loader = await JsonScheduleBackend.loadConferenceFromCfg(this.cfg);
    }

    get conference(): IConference {
        return this.loader.conference;
    };

    get talks(): Map<TalkId, ITalk> {
        return this.loader.talks;
    }

    get auditoriums(): Map<AuditoriumId, IAuditorium> {
        return this.loader.auditoriums;
    }

    get interestRooms(): Map<InterestId, IInterestRoom> {
        return this.loader.interestRooms;
    }
}