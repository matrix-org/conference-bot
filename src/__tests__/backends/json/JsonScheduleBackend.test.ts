import { test, expect, afterEach, beforeEach, describe } from "@jest/globals";
import { Server, createServer } from "node:http";
import { AddressInfo } from "node:net";
import path from "node:path";
import * as fs from "fs";
import { IConfig, JsonScheduleFormat } from "../../../config";
import { JsonScheduleBackend } from "../../../backends/json/JsonScheduleBackend";

function getFixture(fixtureFile: string) {
  return fs.readFileSync(path.join(__dirname, fixtureFile), "utf8");
}

function jsonScheduleServer() {
  return new Promise<Server>((resolve) => {
    const server = createServer((req, res) => {
      if (req.url === "/schedule.json") {
        res.writeHead(200);
        const json = getFixture("original_democon.json");
        res.end(json);
      } else if (req.url === "/fosdem/p/matrix") {
        res.writeHead(200);
        const json = getFixture("fosdem_democon.json");
        res.end(json);
      } else {
        console.log(req.url);
        res.writeHead(404);
        res.end("Not found");
      }
    }).listen(undefined, "127.0.0.1", undefined, () => {
      resolve(server);
    });
  });
}

describe("JsonScheduleBackend", () => {
  let serv;
  beforeEach(async () => {
    serv = await jsonScheduleServer();
  });
  afterEach(async () => {
    serv.close();
  });

  test("can parse a FOSDEM JSON format", async () => {
    const globalConfig = { conference: { name: "DemoCon" } } as IConfig;
    const backend = await JsonScheduleBackend.new(
      "/dev/null",
      {
        backend: "json",
        scheduleFormat: JsonScheduleFormat.FOSDEM,
        scheduleDefinition: `http://127.0.0.1:${
          (serv.address() as AddressInfo).port
        }/fosdem/p/matrix`,
      },
      globalConfig
    );
    expect(backend.conference.title).toBe("DemoCon");
    expect(backend.auditoriums).toMatchSnapshot("auditoriums");
    expect(backend.talks).toMatchSnapshot("talks");
    expect(backend.interestRooms.size).toBe(0);
  });
});
