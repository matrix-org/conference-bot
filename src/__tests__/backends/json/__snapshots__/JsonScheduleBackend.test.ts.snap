// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonScheduleBackend can parse a FOSDEM JSON format: auditoriums 1`] = `
Map {
  "5" => {
    "extraPeople": [
      {
        "email": "conal@example.org",
        "id": "1",
        "matrix_id": "@conal:example.org",
        "name": "Conal Coordinator",
        "role": "coordinator",
      },
    ],
    "id": "5",
    "isPhysical": true,
    "kind": "auditorium",
    "name": "Open CAD/CAM devroom",
    "slug": "open_cad_cam",
    "talks": Map {
      "3" => {
        "auditoriumId": "5",
        "dateTs": 1706745600000,
        "endTime": 1706778000000,
        "id": "3",
        "livestream_endTime": 1706778000000,
        "prerecorded": false,
        "qa_startTime": 0,
        "speakers": [
          {
            "email": "ioferrell@localhost",
            "id": "4",
            "matrix_id": "",
            "name": "Iarlaith O'Ferrell",
            "role": "speaker",
          },
          {
            "email": "niamh@localhost",
            "id": "5",
            "matrix_id": "",
            "name": "Niamh O'Donnellan",
            "role": "speaker",
          },
          {
            "email": "conal@example.org",
            "id": "1",
            "matrix_id": "@conal:example.org",
            "name": "Conal Cördinator",
            "role": "coordinator",
          },
        ],
        "startTime": 1706778000000,
        "subtitle": "",
        "title": "Panacea3D: The Final Word in Open Source 3D Solid Modelling",
      },
    },
  },
  "6" => {
    "extraPeople": [
      {
        "email": "chloe@example.org",
        "id": "246",
        "matrix_id": "@chloe:example.org",
        "name": "Chloé Coordinator",
        "role": "coordinator",
      },
    ],
    "id": "6",
    "isPhysical": true,
    "kind": "auditorium",
    "name": "Self-Hosted Software devroom",
    "slug": "selfhosting",
    "talks": Map {
      "14" => {
        "auditoriumId": "6",
        "dateTs": 1706745600000,
        "endTime": 1706779800000,
        "id": "14",
        "livestream_endTime": 1706779800000,
        "prerecorded": false,
        "qa_startTime": 0,
        "speakers": [
          {
            "email": "tysnr@localhost",
            "id": "42",
            "matrix_id": "",
            "name": "Yasunori Takeda",
            "role": "speaker",
          },
          {
            "email": "chloe@example.org",
            "id": "246",
            "matrix_id": "@chloe:example.org",
            "name": "Chloé Coordinator",
            "role": "coordinator",
          },
        ],
        "startTime": 1706779800000,
        "subtitle": "",
        "title": "A new architecture for Local-First Computing without boundaries",
      },
      "19" => {
        "auditoriumId": "6",
        "dateTs": 1706745600000,
        "endTime": 1706783400000,
        "id": "19",
        "livestream_endTime": 1706783400000,
        "prerecorded": false,
        "qa_startTime": 0,
        "speakers": [
          {
            "email": "vivaldo@localhost",
            "id": "94",
            "matrix_id": "@vivaldo:example.org",
            "name": "Vivaldo Corradetti",
            "role": "speaker",
          },
          {
            "email": "chloe@example.org",
            "id": "246",
            "matrix_id": "@chloe:example.org",
            "name": "Chloé Coordinator",
            "role": "coordinator",
          },
        ],
        "startTime": 1706783400000,
        "subtitle": "",
        "title": "Devising a Blended Federation Topology that reaches Cosmique Perfection",
      },
    },
  },
  "7" => {
    "extraPeople": [
      {
        "email": "bert@example.org",
        "id": "0",
        "matrix_id": "@bert:example.org",
        "name": "Bert Boss",
        "role": "coordinator",
      },
    ],
    "id": "7",
    "isPhysical": true,
    "kind": "auditorium",
    "name": "Keynotes (TALKS NOT ANNOUNCED YET)",
    "slug": "key_notes",
    "talks": Map {},
  },
}
`;

exports[`JsonScheduleBackend can parse a FOSDEM JSON format: talks 1`] = `
Map {
  "3" => {
    "auditoriumId": "5",
    "dateTs": 1706745600000,
    "endTime": 1706778000000,
    "id": "3",
    "livestream_endTime": 1706778000000,
    "prerecorded": false,
    "qa_startTime": 0,
    "speakers": [
      {
        "email": "ioferrell@localhost",
        "id": "4",
        "matrix_id": "",
        "name": "Iarlaith O'Ferrell",
        "role": "speaker",
      },
      {
        "email": "niamh@localhost",
        "id": "5",
        "matrix_id": "",
        "name": "Niamh O'Donnellan",
        "role": "speaker",
      },
      {
        "email": "conal@example.org",
        "id": "1",
        "matrix_id": "@conal:example.org",
        "name": "Conal Cördinator",
        "role": "coordinator",
      },
    ],
    "startTime": 1706778000000,
    "subtitle": "",
    "title": "Panacea3D: The Final Word in Open Source 3D Solid Modelling",
  },
  "14" => {
    "auditoriumId": "6",
    "dateTs": 1706745600000,
    "endTime": 1706779800000,
    "id": "14",
    "livestream_endTime": 1706779800000,
    "prerecorded": false,
    "qa_startTime": 0,
    "speakers": [
      {
        "email": "tysnr@localhost",
        "id": "42",
        "matrix_id": "",
        "name": "Yasunori Takeda",
        "role": "speaker",
      },
      {
        "email": "chloe@example.org",
        "id": "246",
        "matrix_id": "@chloe:example.org",
        "name": "Chloé Coordinator",
        "role": "coordinator",
      },
    ],
    "startTime": 1706779800000,
    "subtitle": "",
    "title": "A new architecture for Local-First Computing without boundaries",
  },
  "19" => {
    "auditoriumId": "6",
    "dateTs": 1706745600000,
    "endTime": 1706783400000,
    "id": "19",
    "livestream_endTime": 1706783400000,
    "prerecorded": false,
    "qa_startTime": 0,
    "speakers": [
      {
        "email": "vivaldo@localhost",
        "id": "94",
        "matrix_id": "@vivaldo:example.org",
        "name": "Vivaldo Corradetti",
        "role": "speaker",
      },
      {
        "email": "chloe@example.org",
        "id": "246",
        "matrix_id": "@chloe:example.org",
        "name": "Chloé Coordinator",
        "role": "coordinator",
      },
    ],
    "startTime": 1706783400000,
    "subtitle": "",
    "title": "Devising a Blended Federation Topology that reaches Cosmique Perfection",
  },
}
`;
